rust/mdb_shard/src/intershard_reference_structs.rs

```diff
@@ -0,0 +1,279 @@
+    pub total_dedup_chunks: u32,
+}
+
+impl IntershardReferenceSequenceEntry {
+    pub fn new<I1: TryInto<u32>>(shard_hash: MerkleHash, total_dedup_chunks: I1) -> Self
+    where
+        <I1 as TryInto<u32>>::Error: std::fmt::Debug,
+    {
+        Self {
+            shard_hash,
+            total_dedup_chunks: total_dedup_chunks.try_into().unwrap_or_default(),
+        }
+    }
+
+    pub fn serialize<W: Write>(
+        &self,
+        writer: &mut W,
+    ) -> std::result::Result<usize, std::io::Error> {
+        let mut buf = [0u8; size_of::<Self>()];
+        {
+            let mut writer_cur = std::io::Cursor::new(&mut buf[..]);
+            let writer = &mut writer_cur;
+
+            write_hash(writer, &self.shard_hash)?;
+            write_u32(writer, self.total_dedup_chunks)?;
+        }
+
+        writer.write_all(&buf[..])?;
+
+        Ok(size_of::<IntershardReferenceSequenceEntry>())
+    }
+
+    pub fn deserialize<R: Read>(reader: &mut R) -> std::result::Result<Self, std::io::Error> {
+        let mut v = [0u8; size_of::<IntershardReferenceSequenceEntry>()];
+        reader.read_exact(&mut v[..])?;
+
+        let mut reader_curs = Cursor::new(&v);
+        let reader = &mut reader_curs;
+
+        Ok(Self {
+            shard_hash: read_hash(reader)?,
+            total_dedup_chunks: read_u32(reader)?,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct IntershardReferenceSequence {
+    pub metadata: IntershardReferenceSequenceHeader,
+    pub entries: Vec<IntershardReferenceSequenceEntry>,
+}
+
+impl IntershardReferenceSequence {
+    /// Construct from an iterator over (hash, count) pairs.
+    pub fn from_counts<C: TryInto<u32> + Copy, I: Iterator<Item = (MerkleHash, C)>>(
+        items: I,
+    ) -> Self
+    where
+        <C as TryInto<u32>>::Error: std::fmt::Debug,
+    {
+        let mut entries: Vec<IntershardReferenceSequenceEntry> =
+            Vec::from_iter(items.map(|(shard_hash, count)| {
+                let total_dedup_hit_count: u32 = count.try_into().unwrap_or(u32::MAX);
+
+                IntershardReferenceSequenceEntry {
+                    shard_hash,
+                    total_dedup_chunks: total_dedup_hit_count,
+                }
+            }));
+
+        entries.sort_unstable_by_key(|e| u64::MAX - e.total_dedup_chunks as u64);
+
+        if entries.len() > INTERSHARD_REFERENCE_SIZE_CAP {
+            entries.resize(INTERSHARD_REFERENCE_SIZE_CAP, Default::default());
+        }
+
+        Self {
+            metadata: IntershardReferenceSequenceHeader::new(entries.len()),
+            entries,
+        }
+    }
+
+    pub fn num_bytes(&self) -> u64 {
+        (size_of::<IntershardReferenceSequenceHeader>()
+            + self.entries.len() * size_of::<IntershardReferenceSequenceEntry>()) as u64
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.entries.is_empty()
+    }
+
+    pub fn merge(self, other: IntershardReferenceSequence) -> Self {
```

Comment:
Is this function used anywhere?

Reply:
Not in this PR; it will be in the dependent PR after this.

Reply:
I see